docs/proposals/ssh-known_hosts.rst

```diff
@@ -0,0 +1,211 @@
+development is the replacement of our programmatic use cases of SSH in favour of
+the in package Go solution, a choice stemming from pain points developing more
+sophisticated utility functions for Tarmak and the desire for improvements in
+control of connections to remote hosts.
+
+During development of this replacement it became clear that proper care must be
+taken during authentication of host public keys during connection and manual
+management of our ``ssh_known_hosts`` cluster file. Our current implementation
+allows OpenSSH to maintain this file however, does not exit with an error if
+public keys do not match due to the flag ``StrictHostKeyChecking`` set to
+``no``. Not only does a miss-match in public keys not cause an error, the
+population of known public keys on different authenticated machines to the same
+EC2 hosts will always use the hosts presented public key, meaning the set of
+public keys could potentially be different for users accessing the same cluster.
+
+Objective
+---------
+
+By implementing stricter enforcement of the ``ssh_known_hosts`` file and passing
+it's management to Tarmak, we can improve the security of SSH connections to
+remote hosts. The key high level points to achieving this is as follows:
+
+ - Disable writes from the OpenSSH command to the ``ssh_known_hosts`` file and
+   enforce strict checking.
+ - Enforce that our in package implementation of SSH connections adheres to this
+   file also.
+ - Collect public keys during instance start up that are then stored, tightly
+   coupled with that host. These keys are able to be used as a source of truth
+   for other authenticated users attempting to connect to remote hosts on the
+   cluster that have empty or an incomplete ``ssh_known_hosts`` file.
+
+Changes
+-------
+
+Firstly, we must restrict the OpenSSH command line tool from editing the
+``ssh_known_hosts`` file and strictly enforce it by updating the generator for the
+``ssh_config`` file. This enables Tarmak to take control of the ``ssh_known_hosts``
+file management.
+
+In order to create a source of truth for each host's public key, each instance
+will have it's public key's attached as a tags, shortly after boot time like the
+following:
+
++-------------------------+---------------------------------------------------------------------------+
+| PublicKey_ssh-ed25519_0 | AAAAC3NzaC1lZDI1NTE5AAAAIE90XYYm6GSDlNGejM+aY5dZEe5vK4XyU++89WdGJcDc==EOF |
++-------------------------+---------------------------------------------------------------------------+
+
+The population of these tags will happen at boot time for all instances,
+regardless of whether they have been created from a direct Terraform apply or
+via an Amazon Auto Scaling Group. At execution time, Wing - present on every
+instance - will invoke an Amazon Lambda function for Instance Tagging. Passed to
+this function will be a collection of the instances public keys, it's Amazon
+identity document and matching PKCS7 document.
```

Comment:
While the instance document provides some assurances that it is an actual instance from AWS we should not done the limitations of that in the proposal:

- There is no cryptographic signature that the public keys come actually from the host that is sending the request
- The lambda function has all the information acting like the instance to someone else with that document (e.g. logging into vault with the same method).  

The instance document could be coming form another instance and